### PR TITLE
Also downcast constraints and statistics when downcasting field types

### PR DESCRIPTION
**Pull request by vinceatbluelabs**
A scenario involving a Pandas dataframe with an object dtype but with integers inside was causing an issue - it looks like the code meant to handle situations like that wasn't complete.  The rest of the code is expecting type-consistent constraints and statistics, but this code wasn't adjusting those to match.

Here's the scenario:

```python
#!/usr/bin/env python3

from records_mover import Session
from records_mover.records import existing_table_handling
import pandas as pd

# Specify which file to upload to which table
database = ...
survey_schema = ...
table_name = "dummy_population_vmb"


def upload_file(db, survey_schema, table_name):
    data = {'Population': [11190846, 1303171035, 207847528]}
    df = pd.DataFrame(data,
                      columns=['Population'])

    df['Population'] = df['Population'].astype("Int64")
    df['Population'] = df['Population'].astype("object")

    session = Session()
    session.set_stream_logging()
    records = session.records

    db_engine = session.get_db_engine(db)

    table_handling = existing_table_handling.ExistingTableHandling.APPEND

    source = session.records.sources.dataframe(df=df)
    target = session.records.targets.table(
        schema_name=survey_schema,
        table_name=table_name,
        db_engine=db_engine,
        existing_table_handling=table_handling,
    )
    results = records.move(source, target)
    print(results)
```

and what was happening:

```
Traceback (most recent call last):
  File "./upload_issue_3.py", line 43, in <module>
    upload_file(database, survey_schema, table_name)
  File "./upload_issue_3.py", line 39, in upload_file
    results = records.move(source, target)
  File "/Users/broz/src/records-mover/records_mover/records/mover.py", line 124, in move
    return move(fileobjs_source, records_target, processing_instructions)
  File "/Users/broz/src/records-mover/records_mover/records/mover.py", line 84, in move
    return records_target.move_from_fileobjs_source(records_source,
  File "/Users/broz/src/records-mover/records_mover/records/targets/table/target.py", line 77, in move_from_fileobjs_source
    return DoMoveFromFileobjsSource(self.prep,
  File "/Users/broz/src/records-mover/records_mover/records/targets/table/move_from_fileobjs_source.py", line 59, in move
    schema_sql = self.schema_sql_for_load(schema_obj, self.records_format, driver)
  File "/Users/broz/src/records-mover/records_mover/records/targets/table/base.py", line 27, in schema_sql_for_load
    return tweaked_records_schema.to_schema_sql(driver,
  File "/Users/broz/src/records-mover/records_mover/records/schema/schema/__init__.py", line 126, in to_schema_sql
    return schema_to_schema_sql(records_schema=self,
  File "/Users/broz/src/records-mover/records_mover/records/schema/schema/sqlalchemy.py", line 20, in schema_to_schema_sql
    columns = [f.to_sqlalchemy_column(driver) for f in records_schema.fields]
  File "/Users/broz/src/records-mover/records_mover/records/schema/schema/sqlalchemy.py", line 20, in <listcomp>
    columns = [f.to_sqlalchemy_column(driver) for f in records_schema.fields]
  File "/Users/broz/src/records-mover/records_mover/records/schema/field/__init__.py", line 157, in to_sqlalchemy_column
    return field_to_sqlalchemy_column(self, driver)
  File "/Users/broz/src/records-mover/records_mover/records/schema/field/sqlalchemy.py", line 200, in field_to_sqlalchemy_column
    type_=field.to_sqlalchemy_type(driver),
  File "/Users/broz/src/records-mover/records_mover/records/schema/field/__init__.py", line 152, in to_sqlalchemy_type
    return field_to_sqlalchemy_type(self, driver)
  File "/Users/broz/src/records-mover/records_mover/records/schema/field/sqlalchemy.py", line 132, in field_to_sqlalchemy_type
    raise SyntaxError(f"Incorrect constraint type in {field.name}: {field.constraints}")
SyntaxError: Incorrect constraint type in Population: RecordsSchemaFieldStringConstraints({'required': False, 'unique': False})
```

### CODE REVIEW

File: tests/component/records/schema/field/test_pandas.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import pandas as pd
+
+
+class TestPandas(unittest.TestCase):
+    def test_refine_field_from_series_more_specific(self) -> None:
+        # This test is designed to break when a new field type is
+        # introduced, so you can add new expectations and make sure
+        # the code handles the new type!
+
+        # for field_type in RECORDS_FIELD_TYPES: # TODO
```
Comment (bluelabsbutler) [inline tests/component/records/schema/field/test_pandas.py:21]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: tests/unit/records/sources/test_dataframes.py [hunk 2]
```diff
@@ -49,7 +49,7 @@ def generate_filename(prefix):
             return f"{prefix}.csv"
 
         mock_target_records_format.generate_filename = generate_filename
-        mock_target_records_schema = mock_RecordsSchema.from_dataframe.return_value
+        mock_target_records_schema = mock_RecordsSchema.from_dataframe.return_value.refine_from_dataframe.return_value
```
Comment (bluelabsbutler) [inline tests/unit/records/sources/test_dataframes.py:52]: E501 line too long (118 > 100 characters)

File: tests/unit/records/sources/test_dataframes.py [hunk 3]
```diff
@@ -124,7 +124,7 @@ def generate_filename(prefix):
 
         mock_target_records_format.generate_filename = generate_filename
 
-        mock_target_records_schema = mock_RecordsSchema.from_dataframe.return_value
+        mock_target_records_schema = mock_RecordsSchema.from_dataframe.return_value.refine_from_dataframe.return_value
```
Comment (bluelabsbutler) [inline tests/unit/records/sources/test_dataframes.py:127]: E501 line too long (118 > 100 characters)

File: tests/component/records/schema/field/test_pandas.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            },
+            'time': {
+                'series': pd.Series([datetime.time(hour=12, minute=0, second=0)]),
+                'constraints_type': RecordsSchemaFieldConstraints,
+                'statistics_type': type(None),
+            },
+            'timetz': {
+                'series': pd.Series([datetime.time(hour=12, minute=0, second=0,
+                                                   tzinfo=pytz.timezone('US/Eastern'))]),
+                # TODO: File issue for backlog about not being able to
```
Comment (bluelabsbutler) [inline tests/component/records/schema/field/test_pandas.py:60]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: tests/component/records/schema/field/test_pandas.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                # derive a time with timezone
+                'expected_field_type': 'time',
+                'constraints_type': RecordsSchemaFieldConstraints,
+                'statistics_type': type(None),
+            },
+            'datetime': {
+                'series': pd.Series([datetime.datetime(2020, 1, 1, hour=12)]),
+                # TODO: Figure out why this is
```
Comment (bluelabsbutler) [inline tests/component/records/schema/field/test_pandas.py:68]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: tests/component/records/schema/field/test_pandas.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                'expected_field_type': 'string',
+                'constraints_type': RecordsSchemaFieldStringConstraints,
+                'statistics_type': RecordsSchemaFieldStringStatistics,
+            },
+            'datetimetz': {
+                'series': pd.Series([datetime.datetime(2020, 1, 1, hour=12,
+                                                       tzinfo=pytz.timezone('US/Eastern'))]),
+                # TODO: Figure out why this is
```
Comment (bluelabsbutler) [inline tests/component/records/schema/field/test_pandas.py:76]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/schema/field/constraints/constraints.py [hunk 7]
```diff
@@ -108,6 +108,34 @@ def from_data(data: Optional['FieldConstraintsDict'],
         else:
             return RecordsSchemaFieldConstraints(required=required, unique=unique)
 
+    def downcast(self, field_type: 'FieldType') -> 'RecordsSchemaFieldConstraints':
+        from .integer import RecordsSchemaFieldIntegerConstraints
+        from .decimal import RecordsSchemaFieldDecimalConstraints
+        required = self.required
+        unique = self.unique
+        constraints: RecordsSchemaFieldConstraints
+        # TODO: Can I refactor this into a method on these
```
Comment (bluelabsbutler) [inline records_mover/records/schema/field/constraints/constraints.py:117]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/schema/field/pandas.py [hunk 8]
```diff
@@ -57,7 +57,23 @@ def refine_field_from_series(field: 'RecordsSchemaField',
...
             if RecordsSchemaField.is_more_specific_type(field_type, field.field_type):
-                field.field_type = field_type
+                if field.constraints is None:
+                    constraints = None
+                else:
+                    constraints = field.constraints.downcast(field_type)
+                field = RecordsSchemaField(name=field.name,
+                                           field_type=field_type,
+                                           constraints=constraints,
+                                           # TODO: Can I get this
```
Comment (bluelabsbutler) [inline records_mover/records/schema/field/pandas.py:67]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/schema/field/pandas.py [hunk 9]
```diff
@@ -55,9 +55,22 @@ def refine_field_from_series(field: 'RecordsSchemaField',
     if unique_python_types.size == 1:
         unique_python_type: Type[Any] = unique_python_types[0]
         field_type = field.python_type_to_field_type(unique_python_type)
+        # TODO: Can this next if statement move to a class?
```
Comment (bluelabsbutler) [inline records_mover/records/schema/field/pandas.py:58]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: metrics/bigfiles_high_water_mark [hunk 10]
```diff
@@ -1 +1 @@
-1015
+1036
```
Comment (vinceatbluelabs) [inline metrics/bigfiles_high_water_mark:1]: This is from `records_mover/records/schema/field/__init__.py` increasing in size.  This is a known issue - there's a lot of stuff in there: https://github.com/bluelabsio/records-mover/issues/110

Given that this bugfix already involved splitting apart our test suites, I'm inclined to leave that in the backlog for now and ask forgiveness for the bigfiles slip.

File: records_mover/records/schema/field/__init__.py [hunk 11]
```diff
@@ -64,9 +64,9 @@ def __init__(self,
     def refine_from_series(self,
                            series: 'Series',
                            total_rows: int,
-                           rows_sampled: int) -> None:
+                           rows_sampled: int) -> 'RecordsSchemaField':
```
Comment (vinceatbluelabs) [inline records_mover/records/schema/field/__init__.py:67]: This previously mutated the series passed in; it now either returns the original or a new object.  Other methods in this chain of calls are changed similarly.

File: records_mover/records/schema/field/__init__.py [hunk 12]
```diff
@@ -114,6 +115,10 @@ def python_type_to_field_type(specific_type: Type[Any]) -> Optional['FieldType']
             datetime.date: 'date',
 
             datetime.time: 'time',
+
+            datetime.datetime: 'datetime',
+
+            pd.Timestamp: 'datetime',
```
Comment (vinceatbluelabs) [inline records_mover/records/schema/field/__init__.py:121]: The component test I added revealed a couple of types we could handle a little better when downcasting dataframes.

File: records_mover/records/schema/field/field_types.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+                    'string',
+                    'boolean',
+                    'date',
+                    'time',
+                    'timetz',
+                    'datetime',
+                    'datetimetz']
+
+# Be sure to add new things below in FieldType, too
+RECORDS_FIELD_TYPES: List[str] = list(get_args(FieldType))  # type: ignore
```
Comment (vinceatbluelabs) [inline records_mover/records/schema/field/field_types.py:16]: I renamed this from `types.py` as mypy doesn't support having a file named that when it's working in the mode it's used in for single-file IDE support.

I also dropped the guards, as we now include both typing_inspect and typing_extensions as dependencies.

File: records_mover/records/schema/field/pandas.py [hunk 14]
```diff
@@ -70,7 +70,8 @@ def refine_field_from_series(field: 'RecordsSchemaField',
             if field.statistics is None:
                 field.statistics = statistics
             elif not isinstance(field.statistics, RecordsSchemaFieldStringStatistics):
-                raise SyntaxError("Did not expect to see existing statistics "
-                                  f"for string type: {field.statistics}")
+                raise ValueError("Did not expect to see existing statistics "
+                                 f"for string type: {field.statistics}")
```
Comment (vinceatbluelabs) [inline records_mover/records/schema/field/pandas.py:74]: SyntaxError has some special rules around it; it's documented that the 'filename' field must be set.  As a result, when it's thrown out of code, the unit tests blow up when they try to read that field

ValueError works fine here, though.

File: records_mover/records/schema/field/sqlalchemy.py [hunk 15]
```diff
@@ -127,6 +127,10 @@ def field_from_sqlalchemy_column(column: Column,
 def field_to_sqlalchemy_type(field: 'RecordsSchemaField',
                              driver: 'DBDriver') -> sqlalchemy.types.TypeEngine:
     if field.field_type == 'integer':
+        if field.constraints and\
+           not isinstance(field.constraints, RecordsSchemaFieldIntegerConstraints):
+            raise ValueError(f"Incorrect constraint type in {field.name}: {field.constraints}")
+
```
Comment (vinceatbluelabs) [inline records_mover/records/schema/field/sqlalchemy.py:133]: We were missing the runtime check before doing a cast() below, which is why folks received a type error in our internal use.

File: tests/component/records/schema/field/test_pandas.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+                                                       tzinfo=pytz.timezone('US/Eastern'))]),
+                # refine_field_from_series() is not smart enough to
+                # distinguish whether the datetime objects inside it all
+                # have timezones or not.
+                'expected_field_type': 'datetime',
+                'constraints_type': RecordsSchemaFieldConstraints,
+                'statistics_type': type(None),
+            }
+        }
+        for field_type in RECORDS_FIELD_TYPES:
```
Comment (vinceatbluelabs) [inline tests/component/records/schema/field/test_pandas.py:83]: Pulling this list from the canonical Python type ensures we're testing each of the valid types.

File: tests/unit/records/schema/field/test_pandas.py [hunk 17]
```diff
@@ -56,61 +54,3 @@ def test_field_from_series(self,
...
-                                 series=series,
-                                 total_rows=mock_total_rows,
-                                 rows_sampled=mock_rows_sampled)
-        mock_RecordsSchemaFieldStringStatistics.\
-            assert_called_with(max_length_bytes=None,
-                               max_length_chars=6,
-                               rows_sampled=mock_rows_sampled,
-                               total_rows=mock_total_rows,)
-        self.assertEqual(mock_field.statistics, mock_statistics)
-        self.assertEqual(mock_field.field_type, 'string')
```
Comment (vinceatbluelabs) [inline tests/unit/records/schema/field/test_pandas.py:116]: These tests were actually non-sensical - you'd never see a more specific type determined from this data.

**Comment by vinceatbluelabs** [comment]
Build failure is awaiting a merge of https://github.com/bluelabsio/records-mover/pull/109

File: records_mover/records/schema/field/__init__.py (hunk 11 above)
Comment (cwegrzyn) [reply records_mover/records/schema/field/__init__.py:67]: I love immutability.

File: records_mover/records/schema/schema/pandas.py [hunk 18]
```diff
@@ -44,8 +47,11 @@ def refine_schema_from_dataframe(records_schema: 'RecordsSchema',
...
-    for field in records_schema.fields:
-        series = sampled_df[field.name]
-        field.refine_from_series(series,
+    fields = [
+        field.refine_from_series(sampled_df[field.name],
                                  total_rows=total_rows,
                                  rows_sampled=rows_sampled)
+        for field in records_schema.fields
+    ]
+    return RecordsSchema(fields=fields,
```
Comment (cwegrzyn) [inline records_mover/records/schema/schema/pandas.py:56]: So much immutability. This PR has made my evening.

File: tests/component/records/test_dataframe_schema_sql_creation.py [hunk 19]
```diff
@@ -0,0 +1,40 @@
+from records_mover.records.sources.dataframes import DataframesRecordsSource
+from records_mover.db.redshift.redshift_db_driver import RedshiftDBDriver
+from records_mover.records.processing_instructions import ProcessingInstructions
+from sqlalchemy_redshift.dialect import RedshiftDialect
+import unittest
+from unittest.mock import Mock
+from pandas import DataFrame
+
+
+class TestDataframeSchemaSqlCreation(unittest.TestCase):
```
Comment (cwegrzyn) [inline tests/component/records/test_dataframe_schema_sql_creation.py:10]: Or is _this_ a regression test? 🤔 😁

**Review by cwegrzyn — APPROVED** [review]
This makes sense to me! :shipit:

File: tests/component/records/test_dataframe_schema_sql_creation.py (hunk 19 above)
Comment (vinceatbluelabs) [reply tests/component/records/test_dataframe_schema_sql_creation.py:10]: Re: regression - let me know if you want to talk more about test suite classification!

Comment (cwegrzyn) [reply tests/component/records/test_dataframe_schema_sql_creation.py:10]: Noooooope, just making a joke :)

**Pull request merged**